ports/stm32/ulpi.c

```diff
@@ -0,0 +1,179 @@
+    __IO uint32_t regval = 0;
+
+    /* disable ULPI_CLK by accessing ULPI_PHY */
+    /* read Vendor ID : (Low, High) 0x24,0x04 for USB3300 */
+    regval = USB_ULPI_Read(0x00);
+    if (regval != 0x24) {
+        __fatal_error("ULPI Error 0x00 != 0x24");
+    }
+
+    regval = USB_ULPI_Read(0x01);
+    if (regval != 0x04) {
+        __fatal_error("ULPI Error 0x01 != 0x04");
+    }
+
+    /* read Product ID */
+    regval = USB_ULPI_Read(0x02);
+    if (regval != 0x07) {
+        __fatal_error("ULPI Error 0x02 != 0x07");
+    }
+
+    regval = USB_ULPI_Read(0x03);
+    if (regval != 0x00) {
+        __fatal_error("ULPI Error 0x03 != 0x00");
+    }
+
+    /* Write to scratch register the pattern 0x55 */
+    USB_ULPI_Write(0x16, 0x55);
+    /* Read to scratch Register and check-it again the written Pattern */
+    regval = USB_ULPI_Read(0x16);
+    if (regval != 0x55) {
+        __fatal_error("ULPI Error 0x16 != 0x55");
+    }
+
+    /* Write to scratch register the pattern 0xAA */
+    USB_ULPI_Write(0x16, 0xAA);
+    /* Read to scratch Register and check-it again the written Pattern */
+    regval = USB_ULPI_Read(0x16);
+    if (regval != 0xAA) {
+        __fatal_error("ULPI Error 0x16 != 0xAA");
+    }
+
+    /* read InterfaceControl reg */
+    regval = USB_ULPI_Read(0x07);
+    /* write InterfaceControl reg,to disable PullUp on stp,
+       to avoid USB_PHY wake up when MCU entering standby */
+    USB_ULPI_Write(0x07, regval | 0x80);
+    /* read InterfaceControl reg */
+    regval = USB_ULPI_Read(0x07);
+    if (regval != 0x80) {
+        __fatal_error("ULPI Error 0x07 != 0x80");
+    }
+
+    /* read FunctionControl reg */
+    regval = USB_ULPI_Read(0x04);
+
+    // Reg 0x40 has a different value if USB is disconnected.
+    if (regval != 0x40 && regval != 0x45) {
+        __fatal_error("ULPI Error 0x04 != 0x40 || 0x45");
+    }
+
+    /* write FunctionControl reg,to put PHY into LowPower mode */
+    USB_ULPI_Write(0x04, regval & (~0x40));
+
+    /* read FunctionControl reg again */
+    regval = USB_ULPI_Read(0x04);
+    if (regval != 0x00) {
+        __fatal_error("ULPI Error 0x04 != 0x00");
+    }
+
+    /* Delay 4 ms */
+    HAL_Delay(10);
+}
+
+/**
+ * @brief  This function wakeup the USB PHY from the Low power mode
+ * @param  None
+ * @retval None
+ */
+void ulpi_leave_low_power(void) {
+    // Configure STP as an output pin.
+    mp_hal_pin_config(MICROPY_HW_USB_HS_ULPI_STP, MP_HAL_PIN_MODE_OUTPUT, MP_HAL_PIN_PULL_NONE, 0);
+    mp_hal_pin_config_speed(MICROPY_HW_USB_HS_ULPI_STP, GPIO_SPEED_FREQ_VERY_HIGH);
+
+    // Configure DIR as an input pin.
+    mp_hal_pin_config(MICROPY_HW_USB_HS_ULPI_DIR, MP_HAL_PIN_MODE_INPUT, MP_HAL_PIN_PULL_NONE, 0);
+    mp_hal_pin_config_speed(MICROPY_HW_USB_HS_ULPI_DIR, GPIO_SPEED_FREQ_VERY_HIGH);
+
+    // Set STP pin high
+    mp_hal_pin_high(MICROPY_HW_USB_HS_ULPI_STP);
+
+    /* Wait for DIR to go low */
+    for (uint32_t ticks = HAL_GetTick();
+         ((HAL_GetTick() - ticks) < 500)
```

Comment:
Are SYSTICK IRQs guaranteed to be running in this function?

Reply:
Both: I don't think HAL_GetTick() will work (it won't change value), and the WFI could delay for a long time.

Reply:
I can remove `WFI` but I need to the ticks, that or delay for at least 500ms just to be sure.

Reply:
How about this ?

```c
// Wait for DIR to go low
if (query_irq() == IRQ_STATE_ENABLED) {
    for (uint32_t ticks = HAL_GetTick();
            ((HAL_GetTick() - ticks) < 500)
            && mp_hal_pin_read(MICROPY_HW_USB_HS_ULPI_DIR);) {
        __WFI();
    }
} else {
    for (uint32_t i=0; i<10; i++, HAL_Delay(50)) {
        if (mp_hal_pin_read(MICROPY_HW_USB_HS_ULPI_DIR) == 0) {
            break;
        }
    }
}
```

Reply:
You could just use that second part and it'd work in all cases, and make the code easier to understand.  Also could make it `HAL_Delay(5)` with a range of 100 in the loop, to make it have less latency.